src/components/Posts/Posts.tsx

```diff
@@ -1,15 +1,77 @@
-import React from 'react';
-import Post from '../Post/Post.tsx';
-import {PostsData } from '../../Api/PostsApi.ts';
-import { PostType } from '../../DataTypes/PostType.ts';
+import React, { useEffect, useState, useRef } from 'react';
+import Post from '../Post/Post';
+import { fetchPosts } from '../../Api/PostsApi';
+import { PostType } from '../../DataTypes/PostType';
 import styles from './Posts.module.css';
 
 const Posts: React.FC = () => {
+  const [posts, setPosts] = useState<PostType[]>([]);
+  const [loading, setLoading] = useState<boolean>(false);
+  const [error, setError] = useState<string | null>(null);
+  const [currentPage, setCurrentPage] = useState<number>(1);
+  const [hasMore, setHasMore] = useState<boolean>(true);
+  const loadingRef = useRef<boolean>(false);
+
+  const loadPosts = async (page: number) => {
+    loadingRef.current = true;
+    setLoading(true);
+    try {
+      const newPosts = await fetchPosts(page);
+      if (newPosts.length === 0) {
+        setHasMore(false);
+      } else {
+        setPosts(prevPosts => [...prevPosts, ...newPosts]);
+        setCurrentPage(page);
+      }
+    } catch (error : any) {
+      setError('Failed to fetch posts.');
+    } finally {
+      setLoading(false);
+      loadingRef.current = false;
+    }
+  };
+
+  useEffect(() => {
+    loadPosts(currentPage);
+  }, []);
+
+  const handleScroll = () => {
+    if (
+      hasMore &&
+      !loadingRef.current &&
```

Comment:
why don't you check based on the loading instead of the loadingRef ?

Reply:
I tried it , but during rapid scrolling, the same page was loaded twice , This happens because updating the loading state causes the component to re-render ,but ref avoids this issue since it doesn't make re-render